network-config-analyzer/MultiLayerPropertiesSet.py

```diff
@@ -36,72 +37,293 @@
+        self.negation = res.negation
+        self.allow_all = res.allow_all
+        return self
+
+    def contained_in(self, other):
+        if other.allow_all or (not self.allow_all and not self.values):  # other allows all or self is empty
+            return True
+        if self.allow_all:  # and not other.allow_all
+            return False
+        # else, both list do not allow all values
+        if not self.negation and not other.negation:
+            return set(self.values).issubset(other.values)
+        if not self.negation and other.negation:
+            return set(self.values).isdisjoint(other.values)
+        if self.negation and not other.negation:
+            return False
+        # else, self.negation and other.negation:
+        return set(other.values).issubset(self.values)
+
+    def set_attributes(self, attrs, not_attrs):
+        assert (bool(attrs) != bool(not_attrs))  # xor - maybe switch to warning
+        if attrs:
+            self.values = attrs
+            self.negation = False
+            self.allow_all = False
+        else:
+            self.values = not_attrs
+            self.negation = True
+            self.allow_all = False
+
+
 class RequestAttrs:
-    http_methods = {"GET", "HEAD", "POST", "PUT", "DELETE", "CONNECT", "OPTIONS", "TRACE", "PATCH"}
+    http_methods = {'GET', 'HEAD', 'POST', 'PUT', 'DELETE', 'CONNECT', 'OPTIONS', 'TRACE', 'PATCH'}
 
     def __init__(self, allow_all=False):
-        self.methods = set()  # set of strings, for allowed request.method values
+        self.methods = set()  # a set of strings for allowed request.method values
         if allow_all:
             self.methods |= RequestAttrs.http_methods
+        # a set of UnlimitedHttpAttributes for allowed request.url_path values
+        self.paths = UnlimitedHttpAttributes(allow_all)
+        self.hosts = UnlimitedHttpAttributes(allow_all)
 
     def __bool__(self):
-        return bool(self.methods)
+        return bool(self.methods) | \
+               bool(self.paths) | \
+               bool(self.hosts)
 
     def __eq__(self, other):
         if isinstance(other, RequestAttrs):
-            res = self.methods == other.methods
+            res = self.methods == other.methods and \
+                  self.paths == other.paths and \
+                  self.hosts == other.hosts
             return res
         return NotImplemented
 
     def __str__(self):
-        return f"request.method in {str(self.methods)}"
+        return f'request.method in {self.methods}\n' \
+               f'request.url_path {self.paths}\n' \
+               f'request.hosts {self.hosts}\n'
 
     def copy(self):
         res = RequestAttrs()
-        res.methods |= self.methods
+        res.methods = self.methods.copy()
+        res.paths = self.paths.copy()
+        res.hosts = self.hosts.copy()
         return res
 
     def __and__(self, other):
         res = RequestAttrs()
-        res.methods = self.methods & other.methods
```

Comment:
why not using the `&` operator for methods?

Reply:
changed for all operators